File: Array/3Sum With Multiplicity/solution.ts
function threeSumMulti(arr: number[], target: number): number {
    let freq = new Map()
    let result = 0
    const MOD = 1e9 + 7
    for(let num of arr) {
        freq.set(num, (freq.get(num) || 0) + 1)
    }
    for(let i = 0; i < arr.length - 2; i++) {
        freq.set(arr[i], freq.get(arr[i]) - 1)
        for(let j = i + 1; j < arr.length - 1; j++) {  
            freq.set(arr[j], freq.get(arr[j]) - 1)
            result += freq.get(target - arr[i] - arr[j]) || 0   
            result = result % MOD
        }
        
        for(let j = i + 1; j < arr.length - 1; j++) {
            freq.set(arr[j], freq.get(arr[j]) + 1)
        }
    }
    return result
};

File: Array/3Sum With Multiplicity/solution.py
class Solution:
    def threeSumMulti(self, arr: List[int], target: int) -> int:
        counter = Counter(arr)
        n = len(arr)
        result = 0
        
        for i in counter.keys():
            for j in counter.keys():
                k = target - i - j
                if k not in counter:
                    continue
                if i == j and j == k:
                    result += counter[i] * (counter[j] - 1) * (counter[k] - 2) // 6
                elif i == j:
                    result += (counter[i] - 1) * counter[j] * counter[k] // 2
                elif i < j < k:
                    result += counter[i] * counter[j] * counter[k]
        return result % (10**9+7)